src/lib/environment.ts

```diff
@@ -1,145 +1,74 @@
-
-    return { childEnv, cancellationToken };
-  }
-
-  public select = <T, P extends any[]>(selector: (state: StateT, ...p1: P) => T, ...args: P): T => {
-    if (this.cancellationToken && this.cancellationToken.canceled) {
-      throw new SagaCancelledError(`Saga has been cancelled`);
-    }
-
-    return selector(this.store.getState(), ...args);
-  };
-
-  /**
-   * Calls `f` with the supplied params
-   *
-   * In case the return type is not correctly inferred, wrap the function in an BoundEffect
-   *
-   * @param f
-   * @param params
-   */
-  public call = <T, P extends any[]>(f: (...params: P) => T, ...params: P): T => {
-    if (this.cancellationToken && this.cancellationToken.canceled) {
-      throw new SagaCancelledError(`Saga has been cancelled`);
-    }
-
-    return f(...params);
-  };
-
-  public take = async <Payload>(actionCreator: ActionCreator<Payload>): Promise<Payload> => {
-    return (await this.waitForMessage(actionCreator)).payload;
-  };
-
-  /**
-   * Runs the given saga as an attached child.
-   * Cancelling the parent will also cancel the child at the next opportunity.
-   *
-   * @param effectOrEffectCreator
-   * @param params
-   */
-  public run<T, P extends any[]>(effect: BoundEffect<Environment<StateT, ActionT>, P, T>): T;
-  public run<T, P extends any[]>(effect: SagaFunc<StateT, ActionT, P, T>, ...params: P): T;
-  public run<P extends any[], T>(
-    effectOrEffectCreator: BoundEffect<Environment<StateT, ActionT>, P, T> | SagaFunc<StateT, ActionT, P, T>,
-    ...args: P
-  ): T {
-    // TODO: Add overload for (env: Env, ...) => T
-    // Can we get rid of `withEnv` then?
-
-    if (effectOrEffectCreator instanceof BoundEffect) {
-      return effectOrEffectCreator.run(this, ...effectOrEffectCreator.args);
-    } else {
-      return effectOrEffectCreator(this, ...args);
-    }
-  }
-
-  /**
-   * Spawns the saga in a new context, returning a detached task
-   *
-   * Cancelling this `Task` will not cancel the parent.
-   *
-   * @param effect
-   * @param params
-   */
-  public spawn<T, P extends any[]>(effect: BoundEffect<Environment<StateT, ActionT>, P, T>): Task<T>;
-  public spawn<T, P extends any[]>(effect: SagaFunc<StateT, ActionT, P, T>, ...params: P): Task<T>;
-  public spawn<P extends any[], T>(
-    effectOrEffectCreator: BoundEffect<Environment<StateT, ActionT>, P, T> | SagaFunc<StateT, ActionT, P, T>,
-    ...args: P
-  ): Task<T> {
-    const { childEnv, cancellationToken } = this.createDetachedChildEnvironment();
-
-    const task: Task<T> = {
-      cancel: () => cancellationToken.cancel(),
-      result:
-        effectOrEffectCreator instanceof BoundEffect
+import { SagaEnvironment, WaitForAction, BoundEffect, FuncWithEnv } from './types';
+
+export function createSagaEnvironment<State>(
+  store: MiddlewareAPI<any, State>,
+  waitForAction: WaitForAction,
+  cancellationToken?: CancellationToken,
+): SagaEnvironment<State> {
```

Comment:
agreed, the `Action` typing isn't really needed in our cases.

I worked on a Redux app a while back that had an explicit type for the action supported by each reducer, and then it would be nice to easily find unused actions because they don't find the type. But since the store is shared with all other "users"/libs, we'd just be fooling ourself assuming anything about the actions.